File: server/internal/ws/crdt_test.go
package ws

import (
	"testing"

	"github.com/google/uuid"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestLWWRegister_Merge(t *testing.T) {
	nodeA := uuid.MustParse("aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa")
	nodeB := uuid.MustParse("bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb")

	t.Run("higher timestamp wins", func(t *testing.T) {
		reg := LWWRegister{Value: "old", Timestamp: 100, NodeID: nodeA}
		accepted := reg.Merge(LWWRegister{Value: "new", Timestamp: 200, NodeID: nodeB})
		assert.True(t, accepted)
		assert.Equal(t, "new", reg.Value)
		assert.Equal(t, int64(200), reg.Timestamp)
	})

	t.Run("lower timestamp loses", func(t *testing.T) {
		reg := LWWRegister{Value: "current", Timestamp: 200, NodeID: nodeA}
		accepted := reg.Merge(LWWRegister{Value: "old", Timestamp: 100, NodeID: nodeB})
		assert.False(t, accepted)
		assert.Equal(t, "current", reg.Value)
	})

	t.Run("same timestamp higher node ID wins", func(t *testing.T) {
		reg := LWWRegister{Value: "a-value", Timestamp: 100, NodeID: nodeA}
		accepted := reg.Merge(LWWRegister{Value: "b-value", Timestamp: 100, NodeID: nodeB})
		assert.True(t, accepted)
		assert.Equal(t, "b-value", reg.Value)
	})

	t.Run("same timestamp lower node ID loses", func(t *testing.T) {
		reg := LWWRegister{Value: "b-value", Timestamp: 100, NodeID: nodeB}
		accepted := reg.Merge(LWWRegister{Value: "a-value", Timestamp: 100, NodeID: nodeA})
		assert.False(t, accepted)
		assert.Equal(t, "b-value", reg.Value)
	})
}

func TestDocumentCRDT_ApplyUpdate(t *testing.T) {
	docID := uuid.New()
	blockID := uuid.New()
	nodeA := uuid.New()
	nodeB := uuid.New()

	t.Run("first update is accepted", func(t *testing.T) {
		crdt := NewDocumentCRDT(docID)
		accepted := crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID,
			BlockID:    blockID,
			Field:      "content",
			Value:      "hello",
			Timestamp:  100,
			NodeID:     nodeA,
		})
		assert.True(t, accepted)
		state := crdt.GetBlockState(blockID)
		require.NotNil(t, state)
		assert.Equal(t, "hello", state.Content.Value)
	})

	t.Run("later update wins over earlier", func(t *testing.T) {
		crdt := NewDocumentCRDT(docID)
		crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID, BlockID: blockID, Field: "content",
			Value: "first", Timestamp: 100, NodeID: nodeA,
		})
		accepted := crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID, BlockID: blockID, Field: "content",
			Value: "second", Timestamp: 200, NodeID: nodeB,
		})
		assert.True(t, accepted)
		assert.Equal(t, "second", crdt.GetBlockState(blockID).Content.Value)
	})

	t.Run("earlier update loses to later", func(t *testing.T) {
		crdt := NewDocumentCRDT(docID)
		crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID, BlockID: blockID, Field: "content",
			Value: "later", Timestamp: 200, NodeID: nodeA,
		})
		accepted := crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID, BlockID: blockID, Field: "content",
			Value: "earlier", Timestamp: 100, NodeID: nodeB,
		})
		assert.False(t, accepted)
		assert.Equal(t, "later", crdt.GetBlockState(blockID).Content.Value)
	})

	t.Run("update rejected on deleted block", func(t *testing.T) {
		crdt := NewDocumentCRDT(docID)
		crdt.ApplyDelete(CRDTDeleteEvent{
			DocumentID: docID, BlockID: blockID, Timestamp: 100, NodeID: nodeA,
		})
		accepted := crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID, BlockID: blockID, Field: "content",
			Value: "revive", Timestamp: 200, NodeID: nodeB,
		})
		assert.False(t, accepted)
	})

	t.Run("concurrent edits different blocks", func(t *testing.T) {
		crdt := NewDocumentCRDT(docID)
		block1 := uuid.New()
		block2 := uuid.New()
		crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID, BlockID: block1, Field: "content",
			Value: "block1-by-A", Timestamp: 100, NodeID: nodeA,
		})
		crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID, BlockID: block2, Field: "content",
			Value: "block2-by-B", Timestamp: 100, NodeID: nodeB,
		})
		assert.Equal(t, "block1-by-A", crdt.GetBlockState(block1).Content.Value)
		assert.Equal(t, "block2-by-B", crdt.GetBlockState(block2).Content.Value)
	})

	t.Run("checked field update", func(t *testing.T) {
		crdt := NewDocumentCRDT(docID)
		accepted := crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID, BlockID: blockID, Field: "checked",
			Value: "true", Timestamp: 100, NodeID: nodeA,
		})
		assert.True(t, accepted)
		assert.Equal(t, "true", crdt.GetBlockState(blockID).Checked.Value)
	})

	t.Run("unknown field rejected", func(t *testing.T) {
		crdt := NewDocumentCRDT(docID)
		accepted := crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID, BlockID: blockID, Field: "unknown",
			Value: "val", Timestamp: 100, NodeID: nodeA,
		})
		assert.False(t, accepted)
	})
}

func TestDocumentCRDT_ApplyDelete(t *testing.T) {
	docID := uuid.New()
	blockID := uuid.New()
	nodeA := uuid.New()

	t.Run("delete marks block as deleted", func(t *testing.T) {
		crdt := NewDocumentCRDT(docID)
		crdt.ApplyUpdate(CRDTUpdateEvent{
			DocumentID: docID, BlockID: blockID, Field: "content",
			Value: "hello", Timestamp: 100, NodeID: nodeA,
		})
		accepted := crdt.ApplyDelete(CRDTDeleteEvent{
			DocumentID: docID, BlockID: blockID, Timestamp: 200, NodeID: nodeA,
		})
		assert.True(t, accepted)
		assert.True(t, crdt.GetBlockState(blockID).Deleted)
	})

	t.Run("duplicate delete is rejected", func(t *testing.T) {
		crdt := NewDocumentCRDT(docID)
		crdt.ApplyDelete(CRDTDeleteEvent{
			DocumentID: docID, BlockID: blockID, Timestamp: 200, NodeID: nodeA,
		})
		accepted := crdt.ApplyDelete(CRDTDeleteEvent{
			DocumentID: docID, BlockID: blockID, Timestamp: 100, NodeID: nodeA,
		})
		assert.False(t, accepted)
	})

	t.Run("delete on nonexistent block creates it deleted", func(t *testing.T) {
		crdt := NewDocumentCRDT(docID)
		newBlock := uuid.New()
		accepted := crdt.ApplyDelete(CRDTDeleteEvent{
			DocumentID: docID, BlockID: newBlock, Timestamp: 100, NodeID: nodeA,
		})
		assert.True(t, accepted)
		state := crdt.GetBlockState(newBlock)
		require.NotNil(t, state)
		assert.True(t, state.Deleted)
	})
}

func TestDocumentCRDT_Clock(t *testing.T) {
	docID := uuid.New()
	crdt := NewDocumentCRDT(docID)

	t.Run("tick advances clock", func(t *testing.T) {
		ts1 := crdt.Tick()
		ts2 := crdt.Tick()
		assert.True(t, ts2 > ts1)
	})

	t.Run("receive tick updates clock", func(t *testing.T) {
		farFuture := int64(9999999999999)
		ts := crdt.ReceiveTick(farFuture)
		assert.True(t, ts > farFuture)
	})

	t.Run("tick else branch when clock is ahead of now", func(t *testing.T) {
		// Set clock far into the future so time.Now().UnixMilli() < Clock
		crdt2 := NewDocumentCRDT(uuid.New())
		crdt2.Clock = int64(99999999999999) // far future
		prev := crdt2.Clock
		ts := crdt2.Tick()
		assert.Equal(t, prev+1, ts, "should increment clock by 1 when now <= clock")
	})
}

func TestDocumentCRDTManager(t *testing.T) {
	manager := NewDocumentCRDTManager()
	docID := uuid.New()

	t.Run("get or create", func(t *testing.T) {
		crdt := manager.GetOrCreate(docID)
		require.NotNil(t, crdt)
		assert.Equal(t, docID, crdt.DocumentID)
		crdt2 := manager.GetOrCreate(docID)
		assert.Equal(t, crdt, crdt2)
	})

	t.Run("has", func(t *testing.T) {
		assert.True(t, manager.Has(docID))
		assert.False(t, manager.Has(uuid.New()))
	})

	t.Run("remove", func(t *testing.T) {
		manager.Remove(docID)
		assert.False(t, manager.Has(docID))
	})
}

File: mobile/src/lib/__tests__/crdt.test.ts
// @ts-nocheck
import {
  LamportClock,
  mergeLWW,
  DocumentCRDT,
  type LWWValue,
  type CRDTUpdateEvent,
} from '../crdt';

describe('LamportClock', () => {
  it('tick advances clock', () => {
    const clock = new LamportClock('node-1');
    const ts1 = clock.tick();
    const ts2 = clock.tick();
    expect(ts2).toBeGreaterThan(ts1);
  });

  it('tick increments when clock is ahead of Date.now', () => {
    const clock = new LamportClock('node-1');
    // Set clock far in the future
    (clock as any).counter = 99999999999999;
    const prev = clock.current;
    const ts = clock.tick();
    expect(ts).toBe(prev + 1);
  });

  it('receiveTick updates to remote if higher', () => {
    const clock = new LamportClock('node-1');
    const remoteTs = 99999999999999;
    const result = clock.receiveTick(remoteTs);
    expect(result).toBe(remoteTs + 1);
  });

  it('receiveTick increments if local is higher', () => {
    const clock = new LamportClock('node-1');
    (clock as any).counter = 99999999999999;
    const prev = clock.current;
    const result = clock.receiveTick(100);
    expect(result).toBe(prev + 1);
  });

  it('current returns counter', () => {
    const clock = new LamportClock('node-1');
    expect(clock.current).toBeGreaterThan(0);
  });

  it('nodeId is set correctly', () => {
    const clock = new LamportClock('my-node');
    expect(clock.nodeId).toBe('my-node');
  });
});

describe('mergeLWW', () => {
  it('remote wins with higher timestamp', () => {
    const local: LWWValue = { value: 'old', timestamp: 100, nodeId: 'a' };
    const remote: LWWValue = { value: 'new', timestamp: 200, nodeId: 'b' };
    expect(mergeLWW(local, remote)).toBe(true);
    expect(local.value).toBe('new');
    expect(local.timestamp).toBe(200);
  });

  it('remote loses with lower timestamp', () => {
    const local: LWWValue = { value: 'current', timestamp: 200, nodeId: 'a' };
    const remote: LWWValue = { value: 'old', timestamp: 100, nodeId: 'b' };
    expect(mergeLWW(local, remote)).toBe(false);
    expect(local.value).toBe('current');
  });

  it('same timestamp higher nodeId wins', () => {
    const local: LWWValue = { value: 'a-val', timestamp: 100, nodeId: 'aaa' };
    const remote: LWWValue = { value: 'b-val', timestamp: 100, nodeId: 'bbb' };
    expect(mergeLWW(local, remote)).toBe(true);
    expect(local.value).toBe('b-val');
  });

  it('same timestamp lower nodeId loses', () => {
    const local: LWWValue = { value: 'b-val', timestamp: 100, nodeId: 'bbb' };
    const remote: LWWValue = { value: 'a-val', timestamp: 100, nodeId: 'aaa' };
    expect(mergeLWW(local, remote)).toBe(false);
    expect(local.value).toBe('b-val');
  });
});

describe('DocumentCRDT', () => {
  const DOC_ID = 'doc-1';
  const NODE_A = 'node-a';
  const NODE_B = 'node-b';

  function makeUpdate(overrides: Partial<CRDTUpdateEvent> = {}): CRDTUpdateEvent {
    return {
      documentId: DOC_ID,
      blockId: 'block-1',
      field: 'content',
      value: 'hello',
      timestamp: 100,
      nodeId: NODE_A,
      ...overrides,
    };
  }

  it('applyRemoteUpdate creates block state', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    const accepted = crdt.applyRemoteUpdate(makeUpdate());
    expect(accepted).toBe(true);
    const state = crdt.getBlockState('block-1');
    expect(state).toBeDefined();
    expect(state?.content.value).toBe('hello');
  });

  it('later update wins', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    crdt.applyRemoteUpdate(makeUpdate({ value: 'first', timestamp: 100 }));
    const accepted = crdt.applyRemoteUpdate(makeUpdate({ value: 'second', timestamp: 200, nodeId: NODE_B }));
    expect(accepted).toBe(true);
    expect(crdt.getBlockState('block-1')?.content.value).toBe('second');
  });

  it('earlier update loses', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    crdt.applyRemoteUpdate(makeUpdate({ value: 'later', timestamp: 200 }));
    const accepted = crdt.applyRemoteUpdate(makeUpdate({ value: 'earlier', timestamp: 100, nodeId: NODE_B }));
    expect(accepted).toBe(false);
    expect(crdt.getBlockState('block-1')?.content.value).toBe('later');
  });

  it('update on deleted block is rejected', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    crdt.applyRemoteDelete(makeUpdate({ action: 'delete', timestamp: 100 }));
    const accepted = crdt.applyRemoteUpdate(makeUpdate({ timestamp: 200 }));
    expect(accepted).toBe(false);
  });

  it('checked field update works', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    const accepted = crdt.applyRemoteUpdate(makeUpdate({ field: 'checked', value: 'true' }));
    expect(accepted).toBe(true);
    expect(crdt.getBlockState('block-1')?.checked.value).toBe('true');
  });

  it('unknown field is rejected', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    const accepted = crdt.applyRemoteUpdate(makeUpdate({ field: 'unknown' }));
    expect(accepted).toBe(false);
  });

  it('applyRemoteDelete marks block as deleted', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    crdt.applyRemoteUpdate(makeUpdate());
    const accepted = crdt.applyRemoteDelete(makeUpdate({ action: 'delete', timestamp: 200 }));
    expect(accepted).toBe(true);
    expect(crdt.getBlockState('block-1')?.deleted).toBe(true);
  });

  it('duplicate delete is rejected', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    crdt.applyRemoteDelete(makeUpdate({ action: 'delete', timestamp: 200 }));
    const accepted = crdt.applyRemoteDelete(makeUpdate({ action: 'delete', timestamp: 100 }));
    expect(accepted).toBe(false);
  });

  it('delete on nonexistent block creates it deleted', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    const accepted = crdt.applyRemoteDelete(makeUpdate({ blockId: 'new-block', action: 'delete' }));
    expect(accepted).toBe(true);
    expect(crdt.getBlockState('new-block')?.deleted).toBe(true);
  });

  it('createUpdateEvent advances clock and tracks state', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    const event = crdt.createUpdateEvent('block-1', 'content', 'edited');
    expect(event.documentId).toBe(DOC_ID);
    expect(event.blockId).toBe('block-1');
    expect(event.field).toBe('content');
    expect(event.value).toBe('edited');
    expect(event.action).toBe('update');
    expect(event.nodeId).toBe(NODE_A);
    expect(event.timestamp).toBeGreaterThan(0);

    const state = crdt.getBlockState('block-1');
    expect(state?.content.value).toBe('edited');
  });

  it('createUpdateEvent for checked field', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    const event = crdt.createUpdateEvent('block-1', 'checked', 'true');
    expect(event.field).toBe('checked');
    const state = crdt.getBlockState('block-1');
    expect(state?.checked.value).toBe('true');
  });

  it('createUpdateEvent for other field does not crash', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    const event = crdt.createUpdateEvent('block-1', 'other', 'val');
    expect(event.field).toBe('other');
  });

  it('createDeleteEvent marks block deleted and returns event', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    const event = crdt.createDeleteEvent('block-1');
    expect(event.action).toBe('delete');
    expect(event.blockId).toBe('block-1');
    expect(crdt.getBlockState('block-1')?.deleted).toBe(true);
  });

  it('createDeleteEvent on nonexistent block creates state', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    crdt.createDeleteEvent('new-block');
    expect(crdt.getBlockState('new-block')?.deleted).toBe(true);
  });

  it('getBlockState returns undefined for unknown block', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    expect(crdt.getBlockState('nonexistent')).toBeUndefined();
  });

  it('clear removes all blocks', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    crdt.applyRemoteUpdate(makeUpdate());
    crdt.clear();
    expect(crdt.getBlockState('block-1')).toBeUndefined();
  });

  it('documentId is set correctly', () => {
    const crdt = new DocumentCRDT(DOC_ID, NODE_A);
    expect(crdt.documentId).toBe(DOC_ID);
  });
});
